### PR TITLE
Add dotnet-watch

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

File: Directory.Build.props
Comment (wli3) [inline Directory.Build.props:25]: why this change? There are multiple level of parallelism in SDK. The existing test run is fully paralleled. At one point, we have to turn of certain level of parallelism to avoid creating too many process at the same time. If you see this is important, could you run some measurements?

File: eng/Versions.props
Comment (wli3) [inline eng/Versions.props:146]: This should not be the first time we use moq, could you unify other projects using moq with this version?

File: sdk.sln
Comment (wli3) [inline sdk.sln:265]: call it BuiltInTools? It is too general

**Review by wli3 — COMMENTED** [review]
just have a quick glance. Could you also add all the files in this PR to the codeowner? Especially every test assets

File: Directory.Build.props
Comment (pranavkm) [reply Directory.Build.props:25]: I was looking at why my tests weren't running in parallel, and this turned out to be the culprit. It looks like this was changed in master so I wanted to see if if was equally safe to do so in this branch: https://github.com/dotnet/sdk/blob/master/Directory.Build.props#L25

File: eng/Versions.props
Comment (pranavkm) [reply eng/Versions.props:146]: There are a couple existing references to it. I updated the reference Microsoft.NET.Sdk.Publish.Tasks.Tests to use this variable. https://github.com/dotnet/sdk/blob/5e440989fcadb67a4f91dc4da993c85c927e4be5/src/Tests/dotnet.Tests/dotnet.Tests.csproj#L148 also has a reference to Moq, but uses an older version. I wasn't sure if this was oversight or intentional, so left it alone.

File: sdk.sln
Comment (pranavkm) [reply sdk.sln:265]: `GlobalTools`? I'm fine with either.

Comment (wli3) [reply sdk.sln:265]: They are not really global tools (shared by many SDKs). I prefer BuiltInTools

File: eng/Versions.props
Comment (wli3) [reply eng/Versions.props:146]: could you try to make them all the same? If it fails to run we can revert it back

File: src/Tests/dotnet-watch.Tests/DotNetWatcherTests.cs
Comment (captainsafia) [inline src/Tests/dotnet-watch.Tests/DotNetWatcherTests.cs:59]: Not sure what's up but I get an exception when running locally here:

```
[xUnit.net 00:00:56.22]     Microsoft.DotNet.Watcher.Tools.DotNetWatcherTests.RunsWithIterationEnvVariable [FAIL]
  Failed Microsoft.DotNet.Watcher.Tools.DotNetWatcherTests.RunsWithIterationEnvVariable [9 s]
  Error Message:
   System.FormatException : Input string was not in a correct format.
  Stack Trace:
     at System.Number.ThrowOverflowOrFormatException(ParsingStatus status, TypeCode type)
   at System.Number.ParseInt32(ReadOnlySpan`1 value, NumberStyles styles, NumberFormatInfo info)
   at System.Int32.Parse(String s, IFormatProvider provider)
   at Microsoft.DotNet.Watcher.Tools.DotNetWatcherTests.RunsWithIterationEnvVariable() in /Users/captainsafia/github.com/dotnet/sdk/src/Tests/dotnet-watch.Tests/DotNetWatcherTests.cs:line 59
--- End of stack trace from previous location ---
```

File: src/Tests/dotnet-watch.Tests/BrowserLaunchTests.cs
Comment (captainsafia) [inline src/Tests/dotnet-watch.Tests/BrowserLaunchTests.cs:40]: Getting an exception locally here. It looks like it doesn't repro on the Darwin build in CI so not sure if it is reproable. Might be a heads up as to the flakiness of this test case.

```
  Failed Microsoft.DotNet.Watcher.Tools.BrowserLaunchTests.LaunchesBrowserOnStart [2 m 9 s]
  Error Message:
   System.Threading.Tasks.TaskCanceledException : A task was canceled.
  Stack Trace:
     at Microsoft.DotNet.Watcher.Tools.AwaitableProcess.GetOutputLineAsync(String predicateName, Predicate`1 predicate, CancellationToken cancellationToken) in /Users/captainsafia/github.com/dotnet/sdk/src/Tests/dotnet-watch.Tests/Utilities/AwaitableProcess.cs:line 94
   at Microsoft.DotNet.Watcher.Tools.AwaitableProcess.GetOutputLineStartsWithAsync(String message, TimeSpan timeout) in /Users/captainsafia/github.com/dotnet/sdk/src/Tests/dotnet-watch.Tests/Utilities/AwaitableProcess.cs:line 87
   at Microsoft.DotNet.Watcher.Tools.BrowserLaunchTests.LaunchesBrowserOnStart() in /Users/captainsafia/github.com/dotnet/sdk/src/Tests/dotnet-watch.Tests/BrowserLaunchTests.cs:line 40
--- End of stack trace from previous location ---
```

File: src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs
Comment (captainsafia) [inline src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:246]: Nice addition!

**Comment by wli3** [comment]
What is the build time and test time impact? Now it is green

File: src/Tests/dotnet-watch.Tests/Utilities/AssertEx.cs
Comment (pranavkm) [reply src/Tests/dotnet-watch.Tests/Utilities/AssertEx.cs:11]: This one's a bit funky. It's comparing two sets of strings, using funky file normalization. Not really files on disk.

File: src/BuiltInTools/dotnet-watch/Internal/MsBuildFileSetFactory.cs
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/Internal/MsBuildFileSetFactory.cs:73]: Yup: https://github.com/dotnet/aspnetcore/issues/8321. I'll remove the TODO

File: src/BuiltInTools/dotnet-watch/Internal/MsBuildFileSetFactory.cs
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/Internal/MsBuildFileSetFactory.cs:165]: I guess it's for users to figure out if the build is being invoked by watch. It's been around since forever

File: src/BuiltInTools/dotnet-watch/dotnet-watch.slnf
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/dotnet-watch.slnf:1]: I kinda wanted a easier way to work with watch specific projects.

**Comment by pranavkm** [comment]
> What is the build time and test time impact

Are there baselines I can compare against? The CI numbers aren't really reliable:

```
Before:
Azure Pipelines / dotnet-sdk-public-ci (Build Darwin Build_Debug)
succeeded 3 days ago in 57m 30s

Azure Pipelines / dotnet-sdk-public-ci (Build Windows_NT Build_Debug)
succeeded 3 days ago in 45m 38s

After:
Azure Pipelines / dotnet-sdk-public-ci (Build Darwin Build_Debug)
succeeded 30 minutes ago in 49m 44s

Azure Pipelines / dotnet-sdk-public-ci (Build Windows_NT Build_Debug)
succeeded 24 minutes ago in 47m 23s
```

**Comment by wli3** [comment]
Test build time locally. Maybe test both clean build and incremental build(make a one line change and build again), with vs without this change. For test time, run this PR several times and see if it still in the existing range.

**Comment by mkArtakMSFT** [comment]
Thanks for working through this folks.
@wli3 there is definitely going to be some impact, given that code is being added here. What is the bar, that will be a red flag for you? I would assume that same bar would be a red flag for us too. Hence, given that @pranavkm didn't notice something already we should be good here.

**Comment by wli3** [comment]
@mkArtakMSFT If it is few seconds, that should be fine(both for clean and incremental build). I don't expect test time will get slower since helix is running in parallel. If it goes over more than that, I hope you can contribute few performance improvements to get it back to what's before. As we identified in the other thread, there are many low hanging fruits.

**Comment by pranavkm** [comment]
Alright, locally build times `build.cmd` are still in the order of seconds:

```
Clean build:
Time Elapsed 00:00:29.15

No-op build:
Time Elapsed 00:00:10.86

Incremental build (change a file in dotnet-watch):
Time Elapsed 00:00:12.47
```

Test times are about ~3m running locally. I suspect that would improve more in `master` where test parallelism is enabled:

```
Passed!  - Failed:     0, Passed:    59, Skipped:     4, Total:    63, Duration: 2 m 47 s - dotnet-watch.Tests.dll (net5.0)
```

File: src/Layout/redist/targets/OverlaySdkOnLKG.targets
Comment (wli3) [inline src/Layout/redist/targets/OverlaySdkOnLKG.targets:44]: There is no good way to test it, but try to make a change in dotnet-watch and exam it reflects in the output of `./build.cmd -pack`. Razor has an issue that the overlay did not really work, and it is still the stage0 over get used and published. And... it is very likely this way now. We assemble dotnet-watch in dotnet/install. So whatever get produced here is likely get overridden https://github.com/dotnet/installer/blob/861a1dd12cb80bd834d0e51442d46ee7d1a4023f/src/redist/targets/BundledDotnetTools.targets

File: src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs
Comment (wli3) [inline src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:66]: It is hard to understand what this is by name

File: src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs
Comment (wli3) [inline src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:64]: Isn't itemspec the "name" of the itemgroup? How to add that in a project?

File: src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj
Comment (wli3) [inline src/BuiltInTools/BrowserRefresh/Microsoft.AspNetCore.Watch.BrowserRefresh.csproj:2]: Entire project finished in sub second, expected? https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-sdk-refs-pull-15188-merge-4ce2d126a25f4ddab9/Microsoft.AspNetCore.Watch.BrowserRefresh.Tests.dll/console.b290de92.log?sv=2019-07-07&se=2021-01-31T19%3A13%3A22Z&sr=c&sp=rl&sig=FFIwC9n2%2BDZEtMSrBHURe9rTcg1Iul2Q%2FbmZx5PkeI8%3D

**Review by wli3 — COMMENTED** [review]
finished a round of review

File: src/Layout/redist/targets/OverlaySdkOnLKG.targets
Comment (pranavkm) [reply src/Layout/redist/targets/OverlaySdkOnLKG.targets:44]: My plan was to continue publishing this as a NuGet package but change the installer to start picking up the package from this repo (instead of AspNetCore). Do you foresee issues with this approach?

File: src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs
Comment (pranavkm) [reply src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs]: Removed.

**Comment by pranavkm** [comment]
@wli3 could we get this in today? There's a few follow up items that I'd like to get merged after this goes in.

File: src/Layout/redist/targets/OverlaySdkOnLKG.targets
Comment (wli3) [reply src/Layout/redist/targets/OverlaySdkOnLKG.targets:44]: I prefer it not to, it is very unintuitive. It would surprise people when debugging. And no SDK component goes a separate route. The fact we have this separate nuget package would be completely due to historical reason.

**Comment by wli3** [comment]
@pranavkm I will approve it now but please create a tracking issue to list out the pending problems before merging it.

**Review by wli3 — APPROVED** [review]

File: src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs
Comment (wli3) [inline src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:64]: What is "ItemElements"? You don't need a separate method for doing it

Comment (pranavkm) [reply src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:64]: Is there a recommended pattern to do this? dotnet-watch has tests that adds items to a project with additional metadata:

```C#
<Compile Include="Blah" Exclude="Baz" Watch="false" />
```

Comment (wli3) [reply src/Tests/Microsoft.NET.TestFramework/ProjectConstruction/TestProject.cs:64]: add more detail to my previous comment. You can use XElement.Parse and just put text in it

**Pull request merged**
